### ISSUE TITLE
Binder + nbgitpuller is broken with latest images

### ISSUE THREAD

**Issue by rabernat** [pangeo-data/pangeo-docker-images#272]
Target repo: https://github.com/rsignell-usgs/hurricane-ike-water-levels

Links generated with https://jupyterhub.github.io/nbgitpuller/link

Working version (Tag from about a month ago): https://binder.pangeo.io/v2/gh/pangeo-data/pangeo-docker-images/2021.10.19?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252Frsignell-usgs%252Fhurricane-ike-water-levels%26urlpath%3Dtree%252Fhurricane-ike-water-levels%252F%26branch%3Dmaster

Broken version (master): https://binder.pangeo.io/v2/gh/pangeo-data/pangeo-docker-images/master?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252Frsignell-usgs%252Fhurricane-ike-water-levels%26urlpath%3Dtree%252Fhurricane-ike-water-levels%252F%26branch%3Dmaster

![image](https://user-images.githubusercontent.com/1197350/142504637-033143e2-21a8-4102-ad02-af1af701d58c.png)

(Link after redirection: `https://hub.binder.pangeo.io/user/pangeo-data-pangeo-docker-images-damrvzsp/git-pull?repo=https%3A%2F%2Fgithub.com%2Frsignell-usgs%2Fhurricane-ike-water-levels&urlpath=tree%2Fhurricane-ike-water-levels%2F&branch=master`)

So something we have done since `2021.10.19` has caused this very common usage to break. 😢

**Comment by scottyhq** [comment]
I'll just not it seems to work on the AWS binder

https://aws-uswest2-binder.pangeo.io/v2/gh/pangeo-data/pangeo-docker-images/master?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252Frsignell-usgs%252Fhurricane-ike-water-levels%26urlpath%3Dtree%252Fhurricane-ike-water-levels%252F%26branch%3Dmaster

So perhaps something isn't working with the binderhub config on GCP? https://github.com/pangeo-data/pangeo-binder

**Comment by rabernat** [comment]
Them I'm going to cc @sgibson91 as our binderhub 🦸‍♀️ . The next step for the Pangeo ➡️  2i2c migration will be the GCP binderhub. Perhaps we can address both issues at the same time.

**Comment by sgibson91** [comment]
I grabbed the logs from trying to use the broken link. Seems there is a "failed login from unknown user" error in the hub pod logs and a problem with not being able to find a runnable browser from the singleuser pod. So maybe there's something up with JupyterLab? A discrepancy between versions on AWS binder and GCP binder that means it works on AWS?

<details> <summary> Logs from the hub pod </summary>

```
[I 2021-11-19 10:28:39.721 JupyterHub log:174] 201 POST /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 29.85ms
[W 2021-11-19 10:28:39.790 JupyterHub base:950] User pangeo-data-pangeo-docker-images-1ams68zu is slow to start (timeout=0)
[I 2021-11-19 10:28:39.790 JupyterHub log:174] 202 POST /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu/servers/ (binder@10.48.18.13) 58.28ms
[I 2021-11-19 10:28:39.817 JupyterHub log:174] 200 GET /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 21.81ms
[I 2021-11-19 10:28:40.833 JupyterHub log:174] 200 GET /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 12.89ms
[I 2021-11-19 10:28:41.949 JupyterHub log:174] 302 GET /user/giswqs-geemap-fnjv4r7h/api/contents/examples/notebooks/15_convert_js_to_py.ipynb?content=0&1637317720885 -> /hub/user/giswqs-geemap-fnjv4r7h/api/contents/examples/notebooks/15_convert_js_to_py.ipynb?content=0&1637317720885 (@177.62.64.140) 2.86ms
[I 2021-11-19 10:28:42.103 JupyterHub log:174] 302 GET /hub/user/giswqs-geemap-fnjv4r7h/api/contents/examples/notebooks/15_convert_js_to_py.ipynb?content=0&1637317720885 -> /hub/login?next=%2Fhub%2Fuser%2Fgiswqs-geemap-fnjv4r7h%2Fapi%2Fcontents%2Fexamples%2Fnotebooks%2F15_convert_js_to_py.ipynb%3Fcontent%3D0%261637317720885 (@177.62.64.140) 3.31ms
[I 2021-11-19 10:28:42.249 JupyterHub log:174] 200 GET /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 10.38ms
[W 2021-11-19 10:28:42.254 JupyterHub base:714] Failed login for unknown user
[W 2021-11-19 10:28:42.256 JupyterHub log:174] 403 GET /hub/login?next=%2Fhub%2Fuser%2Fgiswqs-geemap-fnjv4r7h%2Fapi%2Fcontents%2Fexamples%2Fnotebooks%2F15_convert_js_to_py.ipynb%3Fcontent%3D0%261637317720885 (@177.62.64.140) 4.89ms
[I 2021-11-19 10:28:43.911 JupyterHub log:174] 200 GET /hub/health (@10.48.18.1) 0.98ms
[I 2021-11-19 10:28:44.224 JupyterHub log:174] 200 GET /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 10.66ms
[I 2021-11-19 10:28:46.986 JupyterHub log:174] 200 GET /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 10.63ms
[I 2021-11-19 10:28:50.842 JupyterHub log:174] 200 GET /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 9.85ms
[W 2021-11-19 10:28:53.391 JupyterHub _version:34] Single-user server has no version header, which means it is likely < 0.8. Expected 1.1.0
[I 2021-11-19 10:28:53.391 JupyterHub base:855] User pangeo-data-pangeo-docker-images-1ams68zu took 13.644 seconds to start
[I 2021-11-19 10:28:53.391 JupyterHub proxy:262] Adding user pangeo-data-pangeo-docker-images-1ams68zu to proxy /user/pangeo-data-pangeo-docker-images-1ams68zu/ => http://10.49.9.11:8888
[I 2021-11-19 10:28:53.911 JupyterHub log:174] 200 GET /hub/health (@10.48.18.1) 1.00ms
[I 2021-11-19 10:28:56.241 JupyterHub log:174] 200 GET /hub/api/users/pangeo-data-pangeo-docker-images-1ams68zu (binder@10.48.18.13) 13.20ms
[I 2021-11-19 10:29:01.168 JupyterHub proxy:320] Checking routes
[I 2021-11-19 10:29:03.911 JupyterHub log:174] 200 GET /hub/health (@10.48.18.1) 1.01ms
[I 2021-11-19 10:29:13.911 JupyterHub log:174] 200 GET /hub/health (@10.48.18.1) 1.04ms
```

</details>

<details> <summary> Logs from failed singleuser server </summary>

```
[W 2021-11-19 10:14:17.763 ServerApp] No web browser found: could not locate runnable browser.
ServerApp - ERROR - Uncaught exception GET /user/pangeo-data-pangeo-docker-images-7krze11p/git-pull?repo=https%3A%2F%2Fgithub.com%2Frsignell-usgs%2Fhurricane-ike-water-levels&urlpath=tree%2Fhurricane-ike-water-levels%2F&branch=master&token=K7Qd683gRWmh1-k6rAAqDQ (84.67.31.227)
HTTPServerRequest(protocol='https', host='hub.binder.pangeo.io', method='GET', uri='/user/pangeo-data-pangeo-docker-images-7krze11p/git-pull?repo=https%3A%2F%2Fgithub.com%2Frsignell-usgs%2Fhurricane-ike-water-levels&urlpath=tree%2Fhurricane-ike-water-levels%2F&branch=master&token=K7Qd683gRWmh1-k6rAAqDQ', version='HTTP/1.1', remote_ip='84.67.31.227')
Traceback (most recent call last):
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/tornado/web.py", line 1704, in _execute
    result = await result
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/tornado/gen.py", line 216, in wrapper
    result = ctx_run(func, *args, **kwargs)
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/nbgitpuller/handlers.py", line 175, in get
    self.render_template(
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/notebook/base/handlers.py", line 515, in render_template
    template = self.get_template(name)
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/nbclassic/nbserver.py", line 140, in get_template
    return self.settings['notebook_jinja2_env'].get_template(name)
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/jinja2/environment.py", line 997, in get_template
    return self._load_template(name, globals)
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/jinja2/environment.py", line 958, in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/jinja2/loaders.py", line 125, in load
    source, filename, uptodate = self.get_source(environment, name)
  File "/srv/conda/envs/notebook/lib/python3.9/site-packages/jinja2/loaders.py", line 214, in get_source
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: status.html
[E 2021-11-19 10:14:25.729 ServerApp] {
      "Sec-Gpc": "1",
      "Accept-Language": "en-GB,en-US;q=0.9,en;q=0.8",
      "Accept-Encoding": "gzip, deflate, br",
      "Referer": "https://binder.pangeo.io/",
      "Sec-Fetch-Dest": "document",
      "Sec-Fetch-Mode": "navigate",
      "Sec-Fetch-Site": "same-site",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.69 Safari/537.36",
      "Dnt": "1",
      "Upgrade-Insecure-Requests": "1",
      "Sec-Ch-Ua-Platform": "\"macOS\"",
      "Sec-Ch-Ua-Mobile": "?0",
      "Sec-Ch-Ua": "\"Google Chrome\";v=\"95\", \"Chromium\";v=\"95\", \";Not A Brand\";v=\"99\"",
      "X-Scheme": "https",
      "X-Forwarded-Proto": "https,http",
      "X-Forwarded-Port": "443,80",
      "X-Forwarded-Host": "hub.binder.pangeo.io",
      "X-Forwarded-For": "84.67.31.227,10.48.20.113",
      "X-Real-Ip": "84.67.31.227",
      "X-Request-Id": "9adcd4491f2aa9e00f1ef7a5149c26e2",
      "Host": "hub.binder.pangeo.io",
      "Connection": "close"
    }
[E 2021-11-19 10:14:25.729 ServerApp] 500 GET /user/pangeo-data-pangeo-docker-images-7krze11p/git-pull?repo=https%3A%2F%2Fgithub.com%2Frsignell-usgs%2Fhurricane-ike-water-levels&urlpath=tree%2Fhurricane-ike-water-levels%2F&branch=master&token=K7Qd683gRWmh1-k6rAAqDQ (84.67.31.227) 27.31ms referer=https://binder.pangeo.io/
```

</details>

---

One thing I'm not sure about, and this is because y'all are doing way more advanced stuff with nbgitpuller and the like than me:

There are three environments defined in this repo (all based off one another) and they're defined in subfolders so pulling the default branch actually gets you all three. How do you end up with the "correct" environment?

**Comment by scottyhq** [comment]
> There are three environments defined in this repo (all based off one another) and they're defined in subfolders so pulling the default branch actually gets you all three.

There is a softlink `ln -s pangeo-notebook .binder`, so pangeo-notebook/ defines the environment that is launched. (https://github.com/pangeo-data/pangeo-docker-images/blob/master/.binder)


A key difference between the AWS binder and GCP binder configs currently is that the AWS has auth setup so for example the session maps to `/user/scottyhq` instead of `/user/pangeo-data-pangeo-docker-images-damrvzsp` 

[Helm chart versions](https://github.com/pangeo-data/pangeo-binder/blob/staging/pangeo-binder/requirements.yaml) should be identical unless there were local changes that never were pushed to the repository... In case it's helpful here is the user pod log:

I've never quite understood how jupyterhub chart versions need to match packages in the docker images. For example on AWS the user pod lists `Labels: chart=jupyterhub-0.9.0`

```
 ➜  aws-rtc-12SYJ git:(main) kubectl logs -n prod           jupyter-scottyhq  
[I 2021-11-19 21:43:31.777 SingleUserNotebookApp manager:91] [nb_conda_kernels] enabled, 0 kernels found
[I 2021-11-19 21:43:31.811 SingleUserNotebookApp notebookapp:1593] Authentication of /metrics is OFF, since other authentication is disabled.
[W 2021-11-19 21:43:35.023 LabApp] Config option `kernel_spec_manager_class` not recognized by `LabApp`.
[W 2021-11-19 21:43:35.026 LabApp] Config option `kernel_spec_manager_class` not recognized by `LabApp`.
[W 2021-11-19 21:43:35.032 LabApp] Config option `kernel_spec_manager_class` not recognized by `LabApp`.
[I 2021-11-19 21:43:35.033 LabApp] JupyterLab extension loaded from /srv/conda/envs/notebook/lib/python3.9/site-packages/jupyterlab
[I 2021-11-19 21:43:35.033 LabApp] JupyterLab application directory is /srv/conda/envs/notebook/share/jupyter/lab
generated new fontManager
Patching auth into jupyter_server.base.handlers.JupyterHandler(jupyter_server.base.handlers.AuthenticatedHandler) -> JupyterHandler(jupyterhub.singleuser.mixins.HubAuthenticatedHandler, jupyter_server.base.handlers.AuthenticatedHandler)
[I 2021-11-19 21:43:37.984 SingleUserNotebookApp mixins:576] Starting jupyterhub-singleuser server version 1.5.0
[W 2021-11-19 21:43:37.990 SingleUserNotebookApp _version:70] jupyterhub version 1.1.0 != jupyterhub-singleuser version 1.5.0. This could cause failure to authenticate and result in redirect loops!
[I 2021-11-19 21:43:37.991 SingleUserNotebookApp notebookapp:2302] Serving notebooks from local directory: /home/jovyan
[I 2021-11-19 21:43:37.991 SingleUserNotebookApp notebookapp:2302] Jupyter Notebook 6.4.5 is running at:
```

**Comment by rabernat** [comment]
So I just tried the same like that @scottyhq posted up-thread

https://aws-uswest2-binder.pangeo.io/v2/gh/pangeo-data/pangeo-docker-images/master?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252Frsignell-usgs%252Fhurricane-ike-water-levels%26urlpath%3Dtree%252Fhurricane-ike-water-levels%252F%26branch%3Dmaster

And got the 500 server error! So whatever bug was affecting the GCP binder is not affecting the AWS binder as well. Going back to the `2021.11.09` did work:

https://aws-uswest2-binder.pangeo.io/v2/gh/pangeo-data/pangeo-docker-images/2021.11.09?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252Frsignell-usgs%252Fhurricane-ike-water-levels%26urlpath%3Dtree%252Fhurricane-ike-water-levels%252F%26branch%3Dmaster

I am currently doing a bisection search to figure out the latest tag that does actually work.

**Comment by rabernat** [comment]
✅  `2021.12.02`

I just used the handy `get_versions` script (#290) to look at jupyterhub-singleuser. There was a major version change (1 ➡️  2) in `2022.01.31`. I am guessing this is the root of the problem. 


```
2022.02.04
jupyterhub-singleuser-2.1.1
--
2022.01.31
jupyterhub-singleuser-2.1.1
--
2021.12.02
jupyterhub-singleuser-1.5.0
```

Can anyone see a reason not to pin jupyterhub-singleuser < 2.0 for now? Would that somehow break this image for use with 2i2c JupyterHubs? Pinging @sgibson91 for that question...

**Comment by yuvipanda** [comment]
@rabernat yes, you should definitely pin it to <2.0 for now. I think this means at some point when we upgrade the 2i2c hubs to 2.0 jupyterhub, images will also need to be bumped in a co-ordinated fashion.

**Comment by scottyhq** [comment]
>  you should definitely pin it to <2.0 for now. 

Apologies, looks like I should've gotten some clarity on this conversation before plowing ahead with the version bumps! https://github.com/conda-forge/pangeo-notebook-feedstock/pull/96#pullrequestreview-820701084. 

> I am guessing this is the root of the problem. 

Any way to examine the logs or link an existing issue?

**Comment by yuvipanda** [comment]
Just released nbgitpuller 1.1, which might at least fix some of this.

**Comment by weiji14** [comment]
Closing since Pangeo Binder is not running anymore (#459), and also since we're at nbgitpuller [1.2.0](https://github.com/pangeo-data/pangeo-docker-images/blob/a736179ff9636278175122c669f93afce959637b/pangeo-notebook/packages.txt#L399) as mentioned at https://github.com/pangeo-data/pangeo-docker-images/issues/240#issuecomment-1730774318.
